src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -511,14 +521,153 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol
+                        {
+                            return true;
+                        }
+                        return CheckTupleNamesConstraints(baseType, builder, TupleNamesCheckKind.Deep);
+                    }
+                case TupleNamesCheckKind.Deep:
+                    {
+                        object failedType = type.VisitType((t, diag, _) => !CheckTupleNamesConstraints(t, diag, TupleNamesCheckKind.Shallow), builder);
+                        return failedType == null;
+                    }
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(deepCheckTupleNames);
+            }
+        }
+
+        internal static bool CheckTupleNamesConstraints(TypeSymbol type, Location location, DiagnosticBag diagnostics, TupleNamesCheckKind deepCheckTupleNames)
+        {
+            var builder = ArrayBuilder<TypeParameterDiagnosticInfo>.GetInstance();
+            bool result = CheckTupleNamesConstraints(type, builder, deepCheckTupleNames);
+
+            foreach (var pair in builder)
+            {
+                diagnostics.Add(new CSDiagnostic(pair.DiagnosticInfo, location));
+            }
+            builder.Free();
+
+            return result;
+        }
+
+        /// <summary>
+        /// Checks a single type for tuple name constraints, without recursing into its base types or constituent types.
+        /// </summary>
+        internal static bool CheckTupleNamesConstraintsCore(TypeSymbol type, Location location, DiagnosticBag diagnostics)
+        {
+            var builder = ArrayBuilder<TypeParameterDiagnosticInfo>.GetInstance();
+            bool result = CheckTupleNamesConstraintsCore(type, builder);
+
+            foreach (var pair in builder)
+            {
+                diagnostics.Add(new CSDiagnostic(pair.DiagnosticInfo, location));
+            }
+            builder.Free();
+
+            return result;
+        }
+
+        /// <summary>
+        /// Checks a single type for tuple name constraints, without recursing into its base types or constituent types.
+        /// The tuple name constraints mean that this type may not implement interfaces that are in its base type's AllInterfaces with different tuple names.
+        /// Returns false if there was a problem.
+        /// </summary>
+        private static bool CheckTupleNamesConstraintsCore(TypeSymbol type, ArrayBuilder<TypeParameterDiagnosticInfo> builder)
+        {
+            NamedTypeSymbol baseType = type.BaseTypeNoUseSiteDiagnostics;
+            if ((object)baseType == null)
+            {
+                return true;
+            }
+
+            var interfaces = type.InterfacesAndTheirBaseInterfacesNoUseSiteDiagnostics;
+            if (interfaces.IsEmpty)
+            {
+                return true;
+            }
+
+            var baseInterfaces = new Dictionary<NamedTypeSymbol, NamedTypeSymbol>(TypeSymbol.EqualsIgnoringComparer.InstanceIgnoringTupleNames);
```

Comment:
>var baseInterfaces = new Dictionary<NamedTypeSymbol, NamedTypeSymbol>(TypeSymbol.EqualsIgnoringComparer.InstanceIgnoringTupleNames); [](start = 12, length = 132)

Would it make sense to avoid allocating Dictionary when baseType.AllInterfacesNoUseSiteDiagnostics returns an empty array?